### PR TITLE
#3145 Add IP Address Data Type

### LINKED ISSUE #3145: [FEATURE] Add Support for IP Data Type
**Is your feature request related to a problem?**

OpenSearch SQL plugin does not support the [IP address field type](https://opensearch.org/docs/latest/field-types/supported-field-types/ip/).

🚫 IP address fields are converted to strings:

`search=weblog | fields host`

returns `host` with field type `string`.

🚫 IP address fields cannot be correctly for equality:

`search=weblog | where host = "2001:0db7::ff00:42:8329" | fields host`

will not return the value `2001:0db7:0000:0000:0000:ff00:0042:8329`, even though both strings represent the same IP address.

**What solution would you like?**

Outcomes:
- IP addresses can be retrieved using the OpenSearch SQL plugin without conversion to strings.
- IP addresses supports equality operations (`=` and `!=`).
- IP addresses supports comparison operations (`<`, `<=`, `>`, and `>`) if they are both IPv4 or IPv6.
- IP addresses supports sorting (again, if they are all IPv4 or IPv6).
- IP addresses work with IP-specific functions (currently only `cidrmatch` - see #3110).

Proposed Solution:
- Add an new `IP` type to `ExprCoreType`.
- Replace `OpenSearchExprIpValue` with `ExprIpValue` , and update implementation.
- Update `OpenSearchDataType.MappingType` to map `"ip"` fields to `ExprCoreType.IP`.
- Update `OpenSearchExprValueFactory`.
- Update other code, unit tests, and integration tests as necessary.

**What alternatives have you considered?**

None

**Do you have any additional context?**

This is closely related to https://github.com/opensearch-project/sql/pull/3110. This issue added a new `cidrmatch(ip, cidr)` function that returns whether the given IP address is within the specified CIDR IP address range. As part of this work, the SQL plugin was updated to cast IP addresses to strings - previously, it would raise an exception.

### PR DESCRIPTION
**Pull request by currantw**
Signed-off-by: currantw <taylor.curran@improving.com>

### Description

#3145 Add IP Address Data Type

### Related Issues

Resolves #3145
#3036

### Check List
- [x] New functionality includes testing.
- [x] New functionality has been documented.
 - [x] New functionality has javadoc added.
 - [x] New functionality has a user manual doc added.
- [ ] ~API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).~
- [x] Commits are signed per the DCO using `--signoff`.
- [ ] ~Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/sql/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by YANG-DB — APPROVED** [review]

**Comment by YANG-DB** [comment]
@currantw can we test an IP range query as well?
With operations such as `<` and in range

**Comment by currantw** [comment]
Woops! Didn't mean to close this... 😬

**Comment by YANG-DB** [comment]
@currantw can u plz resolve conflicts ?

**Comment by currantw** [comment]
> @currantw can u plz resolve conflicts ?

Yup. Updated!

**Comment by currantw** [comment]
> @currantw can we test an IP range query as well? With operations such as `<` and in range

I have added a test to `SortCommandIT` for testing sorting of the IP address. There are also unit tests in `ExprIpValueTest` that test `compareTo`.

Let me know which additional tests you think may be needed. I didn't find any existing integration tests for comparisons -- perhaps this is something that should be added?

I'm not sure exactly what you mean by "in range" operations: can you give an example query?

**Review by YANG-DB — APPROVED** [review]

**Comment by YANG-DB** [comment]
@currantw LGTM 👍 
@penghuo @dai-chen @acarbonetto could u plz review ?

**Comment by YANG-DB** [comment]
@currantw can u plz check the failed CI plz ?

File: ppl/src/main/antlr/OpenSearchPPLParser.g4 [hunk 1]
```diff
@@ -363,15 +363,7 @@ wcFieldList
    ;
 
 sortField
-   : (PLUS | MINUS)? sortFieldExpression
-   ;
-
-sortFieldExpression
```
Comment (YANG-DB) [inline ppl/src/main/antlr/OpenSearchPPLParser.g4:369]: @currantw was this `sortFieldExpression` used in a query anywhere ?

**Review by YANG-DB — APPROVED** [review]

File: core/src/main/java/org/opensearch/sql/expression/ip/IPFunctions.java [hunk 2]
```diff
@@ -31,75 +31,35 @@ public void register(BuiltinFunctionRepository repository) {
...
-              "IP address '%s' and CIDR IP address range '%s' are not compatible. Both must be"
-                  + " either IPv4 or IPv6.",
-              addressString, rangeString);
-      throw new SemanticCheckException(msg);
+    if (IPUtils.compare(address, range.getUpper()) > 0) {
+      return ExprValueUtils.LITERAL_FALSE;
     }
 
-    return ExprValueUtils.booleanValue(range.contains(address));
+    return ExprValueUtils.LITERAL_TRUE;
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/expression/ip/IPFunctions.java:63]: nit: 
```
return (IPUtils.compare(address, range.getLower()) < 0) || (IPUtils.compare(address, range.getUpper()) > 0) ? ExprValueUtils.LITERAL_FALSE : ExprValueUtils.LITERAL_TRUE;
```

File: core/src/main/java/org/opensearch/sql/expression/operator/convert/TypeCastOperators.java [hunk 3]
```diff
@@ -173,6 +176,13 @@ private static DefaultFunctionResolver castToBoolean() {
         impl(nullMissingHandling((v) -> v), BOOLEAN, BOOLEAN));
   }
 
+  private static DefaultFunctionResolver castToIp() {
+    return FunctionDSL.define(
+        BuiltinFunctionName.CAST_TO_IP.getName(),
+        impl(nullMissingHandling((v) -> v), IP, IP),
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/expression/operator/convert/TypeCastOperators.java:182]: ordering matters - make sure to but the naive cast last

File: opensearch/src/test/java/org/opensearch/sql/opensearch/data/value/OpenSearchExprValueFactoryTest.java [hunk 4]
```diff
@@ -743,12 +755,14 @@ public void constructStruct() {
...
-    final String valueIp = "192.168.0.1";
+    final String ipString = "192.168.0.1";
     assertEquals(
-        new OpenSearchExprIpValue(valueIp),
-        tupleValue(String.format("{\"%s\":\"%s\"}", fieldIp, valueIp)).get(fieldIp));
+        new ExprIpValue(ipString),
+        tupleValue(String.format("{\"%s\":\"%s\"}", fieldIp, ipString)).get(fieldIp));
   }
 
+  private static final double TOLERANCE = 1E-5;
```
Comment (acarbonetto) [inline opensearch/src/test/java/org/opensearch/sql/opensearch/data/value/OpenSearchExprValueFactoryTest.java:764]: why did you move this?

File: opensearch/src/test/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/RangeQueryTest.java [hunk 5]
```diff
@@ -85,8 +85,8 @@ void test_date_has_format() {
   }
 
   @Test
-  void test_non_date_field_type() {
-    String dateString = "2021-11-08";
+  void test_string_field_type() {
+    String dateString = "STRING";
```
Comment (acarbonetto) [inline opensearch/src/test/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/RangeQueryTest.java:89]: revert?

File: ppl/src/main/antlr/OpenSearchPPLLexer.g4 [hunk 6]
```diff
@@ -52,12 +52,6 @@ WITH:                               'WITH';
 // CLAUSE KEYWORDS
 SORTBY:                             'SORTBY';
 
-// FIELD KEYWORDS
-AUTO:                               'AUTO';
-STR:                                'STR';
-IP:                                 'IP';
-NUM:                                'NUM';
```
Comment (acarbonetto) [inline ppl/src/main/antlr/OpenSearchPPLLexer.g4:59]: revert?

File: ppl/src/main/antlr/OpenSearchPPLParser.g4 (hunk 1 above)
Comment (acarbonetto) [reply ppl/src/main/antlr/OpenSearchPPLParser.g4:369]: it is used in `sortField`, which is used by `Trendline` and `SORT BY`.  @currantw we sure we want to include these changes?

File: ppl/src/main/antlr/OpenSearchPPLLexer.g4 (hunk 6 above)
Comment (acarbonetto) [reply ppl/src/main/antlr/OpenSearchPPLLexer.g4:59]: please add any information about this non-functioning grammar to https://github.com/opensearch-project/sql/issues/3180
We should fix this issue instead of removing the grammar

File: opensearch/src/test/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/RangeQueryTest.java (hunk 5 above)
Comment (currantw) [reply opensearch/src/test/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/RangeQueryTest.java]: Agreed. Done. ✅

File: core/src/main/java/org/opensearch/sql/expression/operator/convert/TypeCastOperators.java (hunk 3 above)
Comment (currantw) [reply core/src/main/java/org/opensearch/sql/expression/operator/convert/TypeCastOperators.java]: Interesting. What is the potential effect of having them in the "wrong" order?

File: core/src/main/java/org/opensearch/sql/expression/ip/IPFunctions.java (hunk 2 above)
Comment (currantw) [reply core/src/main/java/org/opensearch/sql/expression/ip/IPFunctions.java]: Funny you should comment on this: I actually wrote out a comment here, which I apparently forgot to ever actually post 🤦. It went something like:

>  I know that this could be combined into a single statement, but I generally like breaking this out, even if it's more verbose; I tend to find it slightly more readable/understandable. Not sure if there is a style that is more common on this project or in Java? That being said, happy to change it if you think it's better off being more compact

Long way of saying that I've gone ahead and updated it as suggested 😆

File: core/src/main/java/org/opensearch/sql/data/model/ExprIpValue.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+import inet.ipaddr.IPAddress;
+import org.opensearch.sql.data.type.ExprCoreType;
+import org.opensearch.sql.data.type.ExprType;
+import org.opensearch.sql.utils.IPUtils;
+
+/** Expression IP Address Value. */
+public class ExprIpValue extends AbstractExprValue {
+  private final IPAddress value;
+
+  public ExprIpValue(String s) {
```
[the remark this answers is not in the document]
Comment (currantw) [reply core/src/main/java/org/opensearch/sql/data/model/ExprIpValue.java]: Agreed. Updated.

File: ppl/src/main/antlr/OpenSearchPPLLexer.g4 (hunk 6 above)
Comment (currantw) [reply ppl/src/main/antlr/OpenSearchPPLLexer.g4:59]: Reverted and added a `TODO`. `IP` keyword has been moved to the "data type" group - seems to me like it makes the most sense there 🤷

File: ppl/src/main/antlr/OpenSearchPPLParser.g4 (hunk 1 above)
Comment (currantw) [reply ppl/src/main/antlr/OpenSearchPPLParser.g4]: As discussed, I've reverted these changes in favour of addressing this independently in #3180.

File: core/src/main/java/org/opensearch/sql/expression/operator/convert/TypeCastOperators.java (hunk 3 above)
Comment (acarbonetto) [reply core/src/main/java/org/opensearch/sql/expression/operator/convert/TypeCastOperators.java]: There may be an implicit casting on an earlier operation.

File: core/src/main/java/org/opensearch/sql/expression/ip/IPFunctions.java (hunk 2 above)
Comment (acarbonetto) [reply core/src/main/java/org/opensearch/sql/expression/ip/IPFunctions.java]: Oh lol!  Generally, I don't mind verbosity.  But this is too many extra lines of code for my taste.  Generally, I don't have a cut off, but merging 5 lines into 1 sounds like a good tradeoff.

File: DEVELOPER_GUIDE.rst [hunk 8]
```diff
@@ -405,7 +405,7 @@ Sample test class:
 Doctest
 >>>>>>>
 
-Python doctest library makes our document executable which keeps it up-to-date to source code. The doc generator aforementioned served as scaffolding and generated many docs in short time. Now the examples inside is changed to doctest gradually. For more details please read `Doctest <./dev/Doctest.md>`_.
+Python doctest library makes our document executable which keeps it up-to-date to source code. The doc generator aforementioned served as scaffolding and generated many docs in short time. Now the examples inside is changed to doctest gradually. For more details please read `testing-doctest <./docs/dev/testing-doctest.md>`_.
```
Comment (YANG-DB) [inline DEVELOPER_GUIDE.rst:408]: @currantw just making sure - why was changed ?

**Review by YANG-DB — COMMENTED** [review]
@currantw can you plz also add test with `>` , `<` where clause operators ?

File: DEVELOPER_GUIDE.rst (hunk 8 above)
Comment (currantw) [reply DEVELOPER_GUIDE.rst:408]: I noticed that this was incorrect when reading through this documentation. There is no longer a `Doctest.md` file in the code base, so I have updated this to point to the correct file.

File: docs/user/ppl/functions/ip.rst [hunk 9]
```diff
@@ -20,19 +20,19 @@ Argument type: STRING, STRING
...
+    +---------+--------------------+
 
 Note:
- - `ip` can be an IPv4 or an IPv6 address
- - `cidr` can be an IPv4 or an IPv6 block
- - `ip` and `cidr` must be either both IPv4 or both IPv6
- - `ip` and `cidr` must both be valid and non-empty/non-null
+ - `ip` can be an IPv4 or IPv6 address
+ - `cidr` can be an IPv4 or IPv6 block
+ - `ip` and `cidr` must both be valid and non-missing/non-null
```
Comment (currantw) [inline docs/user/ppl/functions/ip.rst:37]: Slight update to the documentation here. Specifically, removed "`ip` and `cidr` must be either both IPv4 or both IPv6", which isn't true (this function works fine on SQL even if the address and block use different IP versions); and changes "non-empty" to "non-missing", which seems to be the more common terminology within the SQL project.

**Review by acarbonetto — APPROVED** [review]
LGTM

File: DEVELOPER_GUIDE.rst (hunk 8 above)
Comment (currantw) [reply DEVELOPER_GUIDE.rst:408]: Noticed while reading through this that `Doctest.md` no longer exists, so corrected the link.

**Comment by currantw** [comment]
> @currantw can you plz also add test with `>` , `<` where clause operators ?

Good call. Added new test class `IPComparisonIT`, which has integration tests for all the comparison operators (i.e. `=`, `!=`, `<`, `<=`, `>`, and `>=`. )

**Review by YANG-DB — APPROVED** [review]
@currantw LGTM !
thanks
